The startup scripts need to output data in the new dict format

At the moment (after the merger of #180) I get this exception:

```
make[1]: Leaving directory '/home/ltratt/warmup_experiment/benchmarks'
python2.7 krun/krun.py startup.krun
[2016-10-05 17:36:24: INFO] Wait 60 secs to allow system to cool prior to collecting initial temperature readings
Traceback (most recent call last):
  File "krun/krun.py", line 346, in <module>
    main(parser)
  File "krun/krun.py", line 209, in main
    inner_main(mailer, config, args)
  File "krun/krun.py", line 295, in inner_main
    sanity_checks(config, platform)
  File "krun/krun.py", line 68, in sanity_checks
    vm_info["vm_def"].sanity_checks()
  File "/home/ltratt/warmup_experiment/krun/krun/vm_defs.py", line 633, in sanity_checks
    self._check_truffle_enabled()
  File "/home/ltratt/warmup_experiment/krun/krun/vm_defs.py", line 629, in _check_truffle_enabled
    "jruby_check_graal_enabled.rb", force_dir=VM_SANITY_CHECKS_DIR)
  File "/home/ltratt/warmup_experiment/krun/krun/util.py", line 250, in spawn_sanity_check
    _ = check_and_parse_execution_results(stdout, stderr, rc)
  File "/home/ltratt/warmup_experiment/krun/krun/util.py", line 216, in check_and_parse_execution_results
    key_set = set(json_data.keys())
AttributeError: 'list' object has no attribute 'keys'
Makefile:58: recipe for target 'bench-startup-no-reboots' failed
make: *** [bench-startup-no-reboots] Error 1
bencher3%
```

--- a/patches/hhvm.diff
+++ b/patches/hhvm.diff
@@ -23,7 +23,7 @@ index 1d0822d..ad3e69b 100644
  # See https://github.com/facebook/hhvm/issues/5217
  INCLUDE(CheckCXXSourceCompiles)
 diff --git a/hphp/runtime/ext/std/ext_std_misc.cpp b/hphp/runtime/ext/std/ext_std_misc.cpp
-index 8b08cec..3347fbf 100644
+index 8b08cec..dcfc57b 100644
 --- a/hphp/runtime/ext/std/ext_std_misc.cpp
 +++ b/hphp/runtime/ext/std/ext_std_misc.cpp
 @@ -44,10 +44,14 @@
@@ -41,13 +41,14 @@ index 8b08cec..3347fbf 100644
  namespace HPHP {
  
  IMPLEMENT_THREAD_LOCAL(std::string, s_misc_highlight_default_string);
-@@ -192,6 +196,14 @@ void StandardExtension::initMisc() {
+@@ -192,6 +196,15 @@ void StandardExtension::initMisc() {
      HHVM_FE(sys_getloadavg);
      HHVM_FE(token_get_all);
      HHVM_FE(token_name);
 +    HHVM_FE(krun_init);
 +    HHVM_FE(krun_done);
 +    HHVM_FE(krun_measure);
++    HHVM_FE(krun_clock_gettime_monotonic);
 +    HHVM_FE(krun_get_num_cores);
 +    HHVM_FE(krun_get_wallclock);
 +    HHVM_FE(krun_get_core_cycles_double);
@@ -56,7 +57,7 @@ index 8b08cec..3347fbf 100644
      HHVM_FE(hphp_to_string);
      HHVM_FALIAS(__SystemLib\\max2, SystemLib_max2);
      HHVM_FALIAS(__SystemLib\\min2, SystemLib_min2);
-@@ -966,5 +978,37 @@ Variant HHVM_FUNCTION(SystemLib_min2, const Variant& value1,
+@@ -966,5 +979,41 @@ Variant HHVM_FUNCTION(SystemLib_min2, const Variant& value1,
  #undef YYTOKEN_MAP
  #undef YYTOKEN
  
@@ -70,6 +71,10 @@ index 8b08cec..3347fbf 100644
 +
 +void HHVM_FUNCTION(krun_measure, int mdata_index) {
 +    krun_measure(mdata_index);
++}
++
++double HHVM_FUNCTION(krun_clock_gettime_monotonic) {
++    return krun_clock_gettime_monotonic();
 +}
 +
 +int HHVM_FUNCTION(krun_get_num_cores) {
@@ -95,16 +100,17 @@ index 8b08cec..3347fbf 100644
  ///////////////////////////////////////////////////////////////////////////////
  }
 diff --git a/hphp/runtime/ext/std/ext_std_misc.h b/hphp/runtime/ext/std/ext_std_misc.h
-index 49db8a5..c0277ea 100644
+index 49db8a5..b3e4d12 100644
 --- a/hphp/runtime/ext/std/ext_std_misc.h
 +++ b/hphp/runtime/ext/std/ext_std_misc.h
-@@ -48,6 +48,14 @@ Variant HHVM_FUNCTION(unpack, const String& format, const String& data);
+@@ -48,6 +48,15 @@ Variant HHVM_FUNCTION(unpack, const String& format, const String& data);
  Array HHVM_FUNCTION(sys_getloadavg);
  Array HHVM_FUNCTION(token_get_all, const String& source);
  String HHVM_FUNCTION(token_name, int64_t token);
 +void HHVM_FUNCTION(krun_init);
 +void HHVM_FUNCTION(krun_done);
 +void HHVM_FUNCTION(krun_measure, int mdata_index);
++double HHVM_FUNCTION(krun_clock_gettime_monotonic);
 +int HHVM_FUNCTION(krun_get_num_cores);
 +double HHVM_FUNCTION(krun_get_wallclock, int mdata_index);
 +double HHVM_FUNCTION(krun_get_core_cycles_double, int mdata_index, int core);
@@ -114,10 +120,10 @@ index 49db8a5..c0277ea 100644
  Variant HHVM_FUNCTION(SystemLib_max2, const Variant& arg1, const Variant& arg2);
  Variant HHVM_FUNCTION(SystemLib_min2, const Variant& arg1, const Variant& arg2);
 diff --git a/hphp/runtime/ext/std/ext_std_misc.php b/hphp/runtime/ext/std/ext_std_misc.php
-index da36140..7b3e11f 100644
+index da36140..461e6e3 100644
 --- a/hphp/runtime/ext/std/ext_std_misc.php
 +++ b/hphp/runtime/ext/std/ext_std_misc.php
-@@ -214,6 +214,30 @@ function token_get_all(string $source): array;
+@@ -214,6 +214,32 @@ function token_get_all(string $source): array;
  <<__ParamCoerceModeFalse, __Native>>
  function token_name(int $token): string;
  
@@ -130,6 +136,8 @@ index da36140..7b3e11f 100644
 +/* Take measurements */
 +<<__Native>>
 +function krun_measure(int $mdata_index): void;
++<<__Native>>
++function krun_clock_gettime_monotonic(): double;
 +
 +/* How many per-core measurements? */
 +<<__Native>>

--- a/patches/v8.diff
+++ b/patches/v8.diff
@@ -1,5 +1,5 @@
 diff --git a/src/d8.cc b/src/d8.cc
-index 0688380..2b1af6c 100644
+index 0688380..728e6ca 100644
 --- a/src/d8.cc
 +++ b/src/d8.cc
 @@ -2,6 +2,9 @@
@@ -37,7 +37,7 @@ index 0688380..2b1af6c 100644
  namespace v8 {
  
  namespace {
-@@ -627,6 +648,56 @@ void Shell::Print(const v8::FunctionCallbackInfo<v8::Value>& args) {
+@@ -627,6 +648,60 @@ void Shell::Print(const v8::FunctionCallbackInfo<v8::Value>& args) {
    fflush(stdout);
  }
  
@@ -64,6 +64,10 @@ index 0688380..2b1af6c 100644
 +    CHECK_NUM_ARGS(args, 1);
 +    DOUBLE_ARG_TO_CINT_CHECKED(0, mdata_index);
 +    krun_measure(mdata_index);
++}
++
++void Shell::KrunClockGettimeMonotonic(const v8::FunctionCallbackInfo<v8::Value>& args) {
++    args.GetReturnValue().Set(krun_clock_gettime_monotonic());
 +}
 +
 +void Shell::KrunGetNumCores(const v8::FunctionCallbackInfo<v8::Value>& args) {
@@ -94,7 +98,7 @@ index 0688380..2b1af6c 100644
  
  void Shell::Write(const v8::FunctionCallbackInfo<v8::Value>& args) {
    for (int i = 0; i < args.Length(); i++) {
-@@ -658,6 +729,33 @@ void Shell::Write(const v8::FunctionCallbackInfo<v8::Value>& args) {
+@@ -658,6 +733,33 @@ void Shell::Write(const v8::FunctionCallbackInfo<v8::Value>& args) {
    }
  }
  
@@ -128,7 +132,7 @@ index 0688380..2b1af6c 100644
  
  void Shell::Read(const v8::FunctionCallbackInfo<v8::Value>& args) {
    String::Utf8Value file(args[0]);
-@@ -1082,10 +1180,50 @@ Local<ObjectTemplate> Shell::CreateGlobalTemplate(Isolate* isolate) {
+@@ -1082,10 +1184,54 @@ Local<ObjectTemplate> Shell::CreateGlobalTemplate(Isolate* isolate) {
            .ToLocalChecked(),
        FunctionTemplate::New(isolate, Print));
    global_template->Set(
@@ -163,6 +167,10 @@ index 0688380..2b1af6c 100644
 +      String::NewFromUtf8(isolate, "krun_measure", NewStringType::kNormal)
 +          .ToLocalChecked(),
 +      FunctionTemplate::New(isolate, KrunMeasure));
++  global_template->Set(
++      String::NewFromUtf8(isolate, "krun_clock_gettime_monotonic", NewStringType::kNormal)
++          .ToLocalChecked(),
++      FunctionTemplate::New(isolate, KrunClockGettimeMonotonic));
 +  global_template->Set(
 +      String::NewFromUtf8(isolate, "krun_get_num_cores", NewStringType::kNormal)
 +          .ToLocalChecked(),
@@ -202,10 +210,10 @@ index f249a78..38a5003 100644
          'd8.h',
          'd8.cc',
 diff --git a/src/d8.h b/src/d8.h
-index 321d9c1..626149f 100644
+index 321d9c1..c63d183 100644
 --- a/src/d8.h
 +++ b/src/d8.h
-@@ -385,7 +385,17 @@ class Shell : public i::AllStatic {
+@@ -385,7 +385,18 @@ class Shell : public i::AllStatic {
                               const  PropertyCallbackInfo<void>& info);
  
    static void Print(const v8::FunctionCallbackInfo<v8::Value>& args);
@@ -217,6 +225,7 @@ index 321d9c1..626149f 100644
 +  static void KrunInit(const v8::FunctionCallbackInfo<v8::Value>& args);
 +  static void KrunDone(const v8::FunctionCallbackInfo<v8::Value>& args);
 +  static void KrunMeasure(const v8::FunctionCallbackInfo<v8::Value>& args);
++  static void KrunClockGettimeMonotonic(const v8::FunctionCallbackInfo<v8::Value>& args);
 +  static void KrunGetNumCores(const v8::FunctionCallbackInfo<v8::Value>& args);
    static void Write(const v8::FunctionCallbackInfo<v8::Value>& args);
 +  static void WriteErr(const v8::FunctionCallbackInfo<v8::Value>& args);

--- a/startup.krun
+++ b/startup.krun
@@ -42,6 +42,7 @@ def patch_runner(vm_def):
         assert "iterations_runner." in vm_def.iterations_runner
         _, suffix = vm_def.iterations_runner.rsplit(".", 1)
         vm_def.iterations_runner = os.path.join(STARTUP_RUNNER_DIR, "startup_runner." + suffix)
+    vm_def.instrument = False
     return vm_def
 
 

--- a/startup_runners/outer_startup_runner.c
+++ b/startup_runners/outer_startup_runner.c
@@ -6,6 +6,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <errno.h>
 #include <stdint.h>
 #include <unistd.h>
@@ -19,18 +20,37 @@ main(int argc, char **argv)
 {
     double start_time = -1;
     int result;
+    char fake_datum[] = "[0, 0], ";
+    char end_datum[] = "[0, 0] ";
+    int datum_size = (int)strlen(fake_datum);
+    krun_init();
+    char* fake_data = malloc(1);
+
+    if (krun_get_num_cores() > 1) {
+        memcpy(fake_data, fake_datum, datum_size);
+    }
+    if (krun_get_num_cores() > 2) {
+        for(int i=0; i < krun_get_num_cores() - 2; i++) {
+            memcpy(fake_data + datum_size * (i + 1), fake_datum, datum_size);
+        }
+    }
+    memcpy(fake_data + datum_size * (krun_get_num_cores() - 1), end_datum, datum_size - 1);
+    fake_data[datum_size * krun_get_num_cores() - 1] = '\0';
 
     /* silence gcc */
     argc = argc;
     argv = argv;
 
-    krun_measure(1);
-    start_time = krun_get_wallclock(1);
-    fprintf(stdout, "[[%f,\n", start_time);
+    start_time = krun_clock_gettime_monotonic();
+    fprintf(stdout, "{ \"core_cycle_counts\": [ %s ],", fake_data);
+    fprintf(stdout, " \"mperf_counts\" : [ %s ],", fake_data);
+    fprintf(stdout, " \"aperf_counts\" : [ %s ],", fake_data);
+    fprintf(stdout, " \"wallclock_times\" : [ %f, ", start_time);
     fflush(stdout);
     result = execv(argv[1], argv + 1);
     if (result) {
         perror("Starting subprocess failed:");
     }
+    free(fake_data);
     return (EXIT_SUCCESS);
 }

--- a/startup_runners/startup_runner.c
+++ b/startup_runners/startup_runner.c
@@ -23,10 +23,8 @@
 int
 main(void)
 {
-    double    start_time = -1;
-
-    krun_measure(1);
-    start_time = krun_get_wallclock(1);
-    fprintf(stdout, "%f], [-1.0, -1.0]]\n", start_time);
+    double start_time = -1;
+    start_time = krun_clock_gettime_monotonic();
+    fprintf(stdout, "%f ] }\n", start_time);
     return (EXIT_SUCCESS);
 }

--- a/startup_runners/startup_runner.java
+++ b/startup_runners/startup_runner.java
@@ -12,9 +12,8 @@ class startup_runner {
     }
 
     public static void main(String args[]) {
-        IterationsRunner.JNI_krun_measure(1);
-        double startTime = IterationsRunner.JNI_krun_get_wallclock(1);
+        double startTime = IterationsRunner.JNI_krun_clock_gettime_monotonic();
         System.out.print(startTime);
-        System.out.println("], [-1.0, -1.0]]");
+        System.out.println("] }");
     }
 }

--- a/startup_runners/startup_runner.js
+++ b/startup_runners/startup_runner.js
@@ -1,3 +1,2 @@
-// NOTE: you need to provide clock_gettime_monotonic.
-print(clock_gettime_monotonic());
-print("], [-1.0, -1.0]]");
+print(krun_clock_gettime_monotonic());
+print("] }");

--- a/startup_runners/startup_runner.lua
+++ b/startup_runners/startup_runner.lua
@@ -1,8 +1,8 @@
 -- Call out to C to get the monotonic time
 local ffi = require("ffi")
-ffi.cdef[[double clock_gettime_monotonic();]]
+ffi.cdef[[ double krun_clock_gettime_monotonic(); ]]
 local kruntime = ffi.load("kruntime")
 
-local BM_start_time = kruntime.clock_gettime_monotonic()
+local BM_start_time = kruntime.krun_clock_gettime_monotonic()
 io.stdout:write(BM_start_time)
-io.stdout:write("], [-1.0, -1.0]]")
+io.stdout:write("] }")

--- a/startup_runners/startup_runner.php
+++ b/startup_runners/startup_runner.php
@@ -1,10 +1,6 @@
 <?php
 
-/*
- * Note! You will need to provide an implementation of clock_gettime_monotonic()
- */
-
-echo clock_gettime_monotonic();
-echo "], [-1.0, -1.0]]";
+echo krun_clock_gettime_monotonic();
+echo "] }";
 
 ?>

--- a/startup_runners/startup_runner.py
+++ b/startup_runners/startup_runner.py
@@ -7,12 +7,10 @@ Executes a benchmark many times within a single process.
 In Kalibera terms, this script represents one executions level run.
 """
 
-import cffi, sys, imp, os
+import cffi
 
 ffi = cffi.FFI()
-ffi.cdef("double clock_gettime_monotonic();")
+ffi.cdef("""double krun_clock_gettime_monotonic(); """)
 libkruntime = ffi.dlopen("libkruntime.so")
 
-clock_gettime_monotonic = libkruntime.clock_gettime_monotonic
-
-print clock_gettime_monotonic(), "], [-1.0, -1.0]]"
+print libkruntime.krun_clock_gettime_monotonic(), "] }"

--- a/startup_runners/startup_runner.rb
+++ b/startup_runners/startup_runner.rb
@@ -1,4 +1,4 @@
-# defined this way so we don't measure the conditional platform check.
+# Defined this way so we don't measure the conditional platform check.
 if /linux/ =~ RUBY_PLATFORM then
     def clock_gettime_monotonic()
         Process.clock_gettime(Process::CLOCK_MONOTONIC_RAW)
@@ -12,5 +12,5 @@ end
 if __FILE__ == $0
     start_time = clock_gettime_monotonic()
     STDOUT.write String(start_time)
-    STDOUT.write "], [-1.0, -1.0]]"
+    STDOUT.write "] }"
 end

